### PR TITLE
Bug 1473356 - Add support for balrog release names

### PR DESCRIPTION
**Pull request by JohanLorenzo**

### CODE REVIEW

File: mozilla_version/firefox.py [hunk 1]
```diff
@@ -43,7 +47,8 @@
 from mozilla_version.version import VersionType
 
 _VALID_VERSION_PATTERN = re.compile(r"""
-^(?P<major_number>\d+)\.(
+^(firefox-)?
```
Comment (tomprince) [inline mozilla_version/firefox.py]: We probably want to look for `devedition` and `thunderbird` here as well.

That being said, I'm wary of parsing all the variant ways of expressing a version willy-nilly. I think we should support parsing both formats, but we should know at the point where we are parsing a version number which style it should be, and should just parse that style.

Comment (JohanLorenzo) [reply mozilla_version/firefox.py]: Sounds good. I'll wait for the proposal you implemented yesterday :)

Comment (JohanLorenzo) [reply mozilla_version/firefox.py]: Done. Latest commits now support `devedition`, `thunderbird`, and `fennec`

**Comment by JohanLorenzo** [comment]
Only the last 3 commits are relevant. Others are from #8 and #9

File: mozilla_version/balrog.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+\.(?P<minor_number>\d+)
+(\.(?P<patch_number>\d+))?
+(
+    (?P<is_nightly>a1)
+    |(?P<is_aurora_or_devedition>a2)
+    |b(?P<beta_number>\d+)
+    |(?P<is_esr>esr)
+)?
+-build(?P<build_number>\d+)
+$""", re.VERBOSE | re.IGNORECASE)
```
Comment (tomprince) [inline mozilla_version/balrog.py]: I don't think you should try to parse the version here. I think you should just capture the entire version string (maybe even with `.*`) and then pass it to the parser for the version. And, if you do that, I don't think you need to make the version parsing logic public.

Comment (JohanLorenzo) [reply mozilla_version/balrog.py]: Good thinking. Fixed!

File: mozilla_version/balrog.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+from mozilla_version.errors import PatternNotMatchedError
+from mozilla_version.parser import get_value_matched_by_regex
+from mozilla_version.gecko import (
+    GeckoVersion, FirefoxVersion, DeveditionVersion, FennecVersion, ThunderbirdVersion
+)
+
+
+_VALID_ENOUGH_BALROG_RELEASE_PATTERN = re.compile(
+    r"^(?P<product>[a-z]+)-(?P<version>.+)$", re.IGNORECASE
```
Comment (tomprince) [inline mozilla_version/balrog.py]: Did you mean to drop `-build?` here?

Comment (tomprince) [reply mozilla_version/balrog.py]: Ah. I guess it is part of the version number. I hadn't realized that the version class parsed build numbers, with a slightly different format. Are there places that have build numbers that aren't balrog slugs?

File: mozilla_version/gecko.py [hunk 4]
```diff
@@ -66,7 +67,7 @@
     |b(?P<beta_number>\d+)
     |(?P<is_esr>esr)
 )?
-(build(?P<build_number>\d+))?$""", re.VERBOSE)
+-?(build(?P<build_number>\d+))?$""", re.VERBOSE)
```
Comment (tomprince) [inline mozilla_version/gecko.py]: I understand this change, and am fine with it.

However, it makes me a little sad that we need to be liberal in what format we expect here.

File: mozilla_version/balrog.py (hunk 3 above)
Comment (JohanLorenzo) [reply mozilla_version/balrog.py]: You're right, it's part of the version number. There are some other places where we deal with build numbers in versions. Ship-it is one example. Some parts [like partials-processing](https://github.com/mozilla-releng/ship-it/blob/18dc35c511c2d38a1f7472c34c435176f3807212/kickoff/views/forms.py#L26) of ship-it handle version numbers like "60.0esrbuild1".

File: mozilla_version/gecko.py (hunk 4 above)
Comment (JohanLorenzo) [reply mozilla_version/gecko.py]: Yeah, it makes me sad too. On the bright side, we're just allowing a dash to be present or not. The data contained in the string isn't changed.

**Pull request merged**